src/snowflake/cli/plugins/nativeapp/artifacts.py

```diff
@@ -75,37 +60,300 @@ class NotInDeployRootError(ClickException):
+             as appropriate.
+            walk_directories (bool): Specifies whether directory to directory mappings should be expanded to
+             resolve their contained files.
+            predicate (PathPredicate): If provided, the predicate is invoked with both the source path and the
+             destination path as arguments. Only pairs selected by the predicate are returned.
+
+        Returns:
+          An iterator over all matching deployed artifacts.
+        """
+        for src in self._src_to_dest.keys():
+            for deployed_src, deployed_dest in self._mappings_for_source(
+                src,
+                absolute=absolute,
+                walk_directories=walk_directories,
+                predicate=predicate,
+            ):
+                yield deployed_src, deployed_dest
+
+    def to_deploy_path(self, src: Path) -> Optional[Path]:
+        """
+        Converts a source path to its corresponding deploy root path. If the input path is relative to the project root,
+        a path relative to the deploy root is returned. If the input path is absolute, an absolute path is returned.
+
+        Returns:
+            The deploy root path for the given source path, or None if no such path exists.
+        """
+        is_absolute = src.is_absolute()
+
+        try:
+            absolute_src = self._absolute_src(src)
+            if not absolute_src.exists():
+                return None
+            canonical_src = self._canonical_src(absolute_src)
+        except ArtifactError:
+            # No mapping is possible for this src path
+            return None
+
+        canonical_dest = self._src_to_dest.get(canonical_src)
+        if canonical_dest is not None:
+            return self._to_output_dest(canonical_dest, is_absolute)
+
+        for canonical_parent in canonical_src.parents:
+            canonical_parent_dest = self.to_deploy_path(canonical_parent)
+            if canonical_parent_dest is not None:
+                return self._to_output_dest(
+                    canonical_parent_dest / canonical_src.relative_to(canonical_parent),
+                    is_absolute,
+                )
+
+        return None
+
+    def _absolute_src(self, src: Path) -> Path:
+        if src.is_absolute():
+            resolved_src = resolve_without_follow(src)
+        else:
+            resolved_src = resolve_without_follow(self._project_root / src)
+        if self._project_root not in resolved_src.parents:
+            raise ArtifactError(
+                f"Source is not in the project root: {src}, root={self._project_root}"
+            )
+        return resolved_src
+
+    def _absolute_dest(self, dest: Path) -> Path:
+        if dest.is_absolute():
+            resolved_dest = resolve_without_follow(dest)
+        else:
+            resolved_dest = resolve_without_follow(self._deploy_root / dest)
+        if (
+            self._deploy_root != resolved_dest
```

Comment:
This is to eliminate condition when src is checked directly under deploy root? In which case dest = deploy_root?

Reply:
This is basically enforcing that any destination path has to be the deploy root or one of its children. The check to make sure we don't clobber the deploy root happens elsewhere, so this logic is good enough here.